### ISSUE TITLE
Conditionals must be updated to match schema

### ISSUE THREAD

**Issue by ajavadia** [Qiskit/qiskit-terra#1283]
From @chriseclectic:

There is a mismatch between how Terra treats conditionals, and how conditionals are expected in the Qobj schema. Specifically, Aer supports conditionals in the new format.

The difference is:

Terra you set a conditional based on the value of an entire classical register represented as an integer: Eg:

```
cr = ClassicalRegister(2)
x(qr[0).c_if(cr, 3)  # conditional for cr = "11"
```

Qobj you set a conditional based on a single "register" bit being 1. Eg:

```
{"name": "x", "qubits": [0], "conditional": [0]}  # apply an x if `register[bit0] = "1"`
{"name": "x", "qubits": [0], "conditional": [1]}  # apply an x if `register[bit1] = "1"`
```

So for single bit classical registers you can just write all measurements to a register field and be done with it. Eg. {"name": "measure", "qubits": [0], "memory": [0], "register": [0]}, but for conditionals on multi-bit ClassicalRegisters you will need to use the boolean function (bfunc).

The bfunc is a qobj instrution: { "name": "bfunc", "mask": "0xF", "relation": "==", "val": "0x5", "register": 0}. This applies a mask to all register bits, and returns a bool for (register AND mask) relation val.

So I think one way to handle this would be to have all measure operations (for simulators) write to both the memory and register, and then enlarge the register by 1 if there are conditionals and have each conditional have a bfunc writing to the additional register location and using that location for the conditional check.

Eg:

```
qr = QuantumRegister(1)
cr = ClassicalRegister(2)
circuit.x(qr[0]).c_if(cr, 0)
circuit.y(qr[0]).c_if(cr, 1)
circuit.z(qr[0]).c_if(cr, 2)
circuit.h(qr[0]).c_if(cr, 3)
```

Compiling to something like this should work:

```
qobj_experiment = {
  "config": {"num_qubits": 1, "memory_slots": 2, "register_slots": 3}
  "instructions": [
    {"name": "bfunc", "relation": "==", "mask": "0x3", "val": "0x0", "register": 2},
    {"name": "x", "qubits": [0], "conditional": [2]},
    {"name": "bfunc", "relation": "==", "mask": "0x3", "val": "0x1", "register": 2},
    {"name": "y", "qubits": [0], "conditional": [2]},
    {"name": "bfunc", "relation": "==", "mask": "0x3", "val": "0x2", "register": 2},
    {"name": "z", "qubits": [0], "conditional": [2]},
    {"name": "bfunc", "relation": "==", "mask": "0x3", "val": "0x3", "register": 2},
    {"name": "h", "qubits": [0], "conditional": [2]}
  ]
}
```

For an experiment you may not have unlimited register slots so you would only want to write to register the measures that are later used for conditionals. So you might want to do that in your implementation.

**Comment by jaygambetta** [comment]
in #1629 I made it work using the old style to pass the test but to move it to the new style we need to discuss if we are going to introduce register_slots yet.

**Comment by jaygambetta** [comment]
can this be done for 0.8

**Comment by ajavadia** [comment]
fixed via #1778
